src/Swarm/TUI/Model.hs

```diff
@@ -387,48 +382,74 @@ lastEntry t h =
 
-mkCmdPrompt :: Text -> REPLPrompt
-mkCmdPrompt t = CmdPrompt t []
+-- | This data type tells us how to interpret the text typed
+--   by the player at the prompt (which is stored in Editor).
+data REPLPrompt
+  = -- | Interpret the prompt text as a regular command.
+    --   The list is for potential completions, which we can
+    --   cycle through by hitting Tab repeatedly
+    CmdPrompt [Text]
+  | -- | Interpret the prompt text as "search this text in history"
+    SearchPrompt REPLHistory
 
 defaultPrompt :: REPLPrompt
-defaultPrompt = mkCmdPrompt ""
+defaultPrompt = CmdPrompt []
+
+data REPLState = REPLState
+  { _replPromptType :: REPLPrompt
+  , _replPromptEditor :: Editor Text Name
+  , _replValid :: Bool
+  , _replLast :: Text
+  , _replType :: Maybe Polytype
+  , _replHistory :: REPLHistory
+  }
 
--- | Lens for accesing the text of the prompt.
---   Notice that setting the text clears any pending completions.
-promptTextL :: Lens' REPLPrompt Text
-promptTextL = lens g s
+newREPLEditor :: Text -> Editor Text Name
+newREPLEditor t = applyEdit gotoEnd $ editorText REPLPanel (Just 1) t
  where
-  -- Notice that the prompt ADT must have a Text field in every constructor (representing what the user writes).
-  -- This should be force in the ADT itself... right know this here
-  -- The compiler will complain about "Non complete patterns" on this two function.
-  g :: REPLPrompt -> Text
-  g (CmdPrompt t _) = t
-  g (SearchPrompt t _) = t
-
-  s :: REPLPrompt -> Text -> REPLPrompt
-  s (CmdPrompt _ _) t = mkCmdPrompt t
-  s (SearchPrompt _ h) t = SearchPrompt t h
-
--- | Turn the repl prompt into a decorator for the form
-replPromptAsWidget :: REPLPrompt -> Widget Name
-replPromptAsWidget (CmdPrompt {}) = txt "> "
-replPromptAsWidget (SearchPrompt t rh) =
-  case lastEntry t rh of
-    Nothing -> txt "[nothing found] "
-    Just lastentry
-      | T.null t -> txt "[find] "
-      | otherwise -> txt $ "[found: \"" <> lastentry <> "\"] "
-
--- | Creates the repl form as a decorated form.
-mkReplForm :: REPLPrompt -> Form REPLPrompt AppEvent Name
-mkReplForm r = newForm [(replPromptAsWidget r <+>) @@= editTextField promptTextL REPLInput (Just 1)] r
+  ls = T.lines t
+  pos = (length ls - 1, T.length (last ls))
+  gotoEnd = if null ls then id else TZ.moveCursor pos
+
+initREPLState :: REPLHistory -> REPLState
+initREPLState = REPLState defaultPrompt (newREPLEditor "") True "" Nothing
+
+makeLensesWith (lensRules & generateSignatures .~ False) ''REPLState
+
+-- | The way we interpret text typed by the player in the REPL prompt.
+replPromptType :: Lens' REPLState REPLPrompt
+
+-- | The prompt where the user can type input at the REPL.
+replPromptEditor :: Lens' REPLState (Editor Text Name)
+
+-- | Convinience lens to get text from editor and replace it with new
+--   one that has the provided text.
+--
+-- TODO: Editor uses a list of Text meaning something? It does not seem to be lines or cursor position.
```

Comment:
What's the status of this TODO?

It looks like `getEditContents` returns the result of `getText` from the `text-zipper` package, which in turn seems to have something to do with returning the part before the focus, the focus, and the part after the focus.  But it looks like just concatenating it is probably fine.

Reply:
I think they may be lines but with the `'\n'` character. Concatenating is the only thing that works.

I will remove the comment and we can investigate it more when implementing multiline REPL.